Category / Tag v2 updates

Miraliu opened the pull request:
1. Add request timeout.
2. Change JSONDecodeError handling since the support for python 3.4 has been dropped.
3. Fix mutable default values from parameters.
4. Refactored BWResources to avoid multiple api calls and cross reference between tags / rules.
5. Updated rule category and rule tags to V2 endpoints

pfairfax1 commented:
Hi @Miraliu , really appreciate you raising this PR. We've decided to use the core part of your code (with minor changes) to support uploading rulecategories and ruletags in a separate PR rather than this one, because this makes quite a few deep changes which would take a while to test and review, and we would prefer these changes be split up into thematically separate PRs, and more clearly defined commits. 

Please do know that we have used parts of your code in [PR-33](https://github.com/BrandwatchLtd/bcr-api/pull/33) and appreciate your input.

Miraliu commented:
> Hi @Miraliu , really appreciate you raising this PR. We've decided to use the core part of your code (with minor changes) to support uploading rulecategories and ruletags in a separate PR rather than this one, because this makes quite a few deep changes which would take a while to test and review, and we would prefer these changes be split up into thematically separate PRs, and more clearly defined commits.
> 
> Please do know that we have used parts of your code in [PR-33](https://github.com/BrandwatchLtd/bcr-api/pull/33) and appreciate your input.
Thanks Peter! Added 1 comment to the new PR, we are current having trouble with categories we updated after migration...
The additional changes were mainly for the circular reference in BWRules that used to need BWCategories / BWTags. Since they are no longer needed there you can ignore most of the changes in BWResources. Just take a look at the error handling changes in BWData / BWProject when you have a chance. Many thanks!!
